src/workerd/api/url-standard.c++

```diff
@@ -27,2093 +27,280 @@
 }
 
-URL::URL(jsg::UsvStringPtr url, jsg::Optional<jsg::UsvStringPtr> base)
-    : inner(handleConstructorParse(url, kj::mv(base))) {}
+URL::URL(kj::StringPtr url, kj::Maybe<kj::StringPtr> base) : inner(parse(url, base)) {}
 
 URL::~URL() noexcept(false) {
   KJ_IF_SOME(searchParams, maybeSearchParams) {
     searchParams->maybeUrl = nullptr;
   }
 }
 
-bool URL::canParse(jsg::UsvString url, jsg::Optional<jsg::UsvString> maybeBase) {
-  KJ_IF_SOME(base, maybeBase) {
-    KJ_IF_SOME(parsedBase, URL::parse(base)) {
-      return URL::parse(url, parsedBase) != kj::none;
-    }
-  }
-  return URL::parse(url) != kj::none;
+bool URL::canParse(kj::String url, jsg::Optional<kj::String> maybeBase) {
+  return jsg::Url::canParse(url, maybeBase.map([](kj::String& str) { return str.asPtr(); }));
 }
 
 jsg::Ref<URLSearchParams> URL::getSearchParams() {
   KJ_IF_SOME(searchParams, maybeSearchParams) {
     return searchParams.addRef();
   }
-  auto searchParams = jsg::alloc<URLSearchParams>(inner.query, *this);
-  maybeSearchParams = searchParams.addRef();
-  return kj::mv(searchParams);
+  return maybeSearchParams.emplace(jsg::alloc<URLSearchParams>(inner.getSearch(), *this)).addRef();
 }
 
-jsg::UsvString URL::getOrigin() {
-  KJ_SWITCH_ONEOF(inner.getOrigin()) {
-    KJ_CASE_ONEOF(opaque, OpaqueOrigin) {
-      return jsg::usv(getCommonStrings().NULL_);
-    }
-    KJ_CASE_ONEOF(tuple, TupleOrigin) {
-      // The additional 9 codepoints here is for the :// and possible port prefix + port.
-      jsg::UsvStringBuilder builder(tuple.scheme.size() + tuple.host.size() + 9);
-      builder.addAll(tuple.scheme);
-      builder.add(':', '/', '/');
-      builder.addAll(tuple.host);
-      KJ_IF_SOME(port, tuple.port) {
-        builder.add(':');
-        builder.addAll(kj::toCharSequence(port));
-      }
-      return builder.finish();
-    }
-  }
-  KJ_UNREACHABLE
+kj::Array<const char> URL::getOrigin() {
+  return inner.getOrigin();
 }
 
-jsg::UsvString URL::getHref() { return inner.getHref(); }
-
-void URL::setHref(jsg::UsvString value) {
-  inner = JSG_REQUIRE_NONNULL(parse(value), TypeError, "Invalid URL string.");
-  KJ_IF_SOME(searchParams, maybeSearchParams) {
-    searchParams->reset(toMaybePtr(inner.query));
-  }
+kj::ArrayPtr<const char> URL::getHref() {
+  return inner.getHref();
 }
 
-jsg::UsvString URL::getProtocol() {
-  jsg::UsvStringBuilder builder(inner.scheme.size() + 1);
-  builder.addAll(inner.scheme);
-  builder.add(':');
-  return builder.finish();
-}
+#define SET(Name, value) inner.set##Name(value)
+#define GET(Name) inner.get##Name();
 
-void URL::setProtocol(jsg::UsvString value) {
-  if (value.empty()) return;
-  jsg::UsvStringBuilder builder(value.size() + 1);
-  builder.addAll(value);
-  builder.add(':');
-  KJ_IF_SOME(record, parse(builder.finish(), nullptr, inner, ParseState::SCHEME_START)) {
-    inner = kj::mv(record);
+
+void URL::setHref(kj::String value) {
+  SET(Href, value);
```

Comment:
I wonder if these are actually worth it. `inner.setHref(value);` is not to terrible either but absolutely explicit and transparent and has no problems with clangd